File: internal/facts/facts.go
package facts

import (
	"sync"
	"time"

	"github.com/huanghe/HomeState/internal/model"
)

// Store holds the raw fact events as a time-series buffer.
type Store struct {
	mu     sync.RWMutex
	events []model.FactEvent
	limit  int
}

func New(limit int) *Store {
	if limit <= 0 {
		limit = 10000
	}
	return &Store{
		events: make([]model.FactEvent, 0, limit),
		limit:  limit,
	}
}

// Record stores a new fact event.
func (s *Store) Record(ev model.FactEvent) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if ev.Timestamp.IsZero() {
		ev.Timestamp = time.Now()
	}
	s.events = append(s.events, ev)
	if len(s.events) > s.limit {
		s.events = s.events[len(s.events)-s.limit:]
	}
}

// Latest returns the most recent event for an entity, or nil.
func (s *Store) Latest(entityID string) *model.FactEvent {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for i := len(s.events) - 1; i >= 0; i-- {
		if s.events[i].EntityID == entityID {
			return &s.events[i]
		}
	}
	return nil
}

// Recent returns the last N events for an entity.
func (s *Store) Recent(entityID string, n int) []model.FactEvent {
	s.mu.RLock()
	defer s.mu.RUnlock()
	var result []model.FactEvent
	for i := len(s.events) - 1; i >= 0 && len(result) < n; i-- {
		if s.events[i].EntityID == entityID {
			result = append([]model.FactEvent{s.events[i]}, result...)
		}
	}
	return result
}

// AllSince returns all events after the given time.
func (s *Store) AllSince(since time.Time) []model.FactEvent {
	s.mu.RLock()
	defer s.mu.RUnlock()
	var result []model.FactEvent
	for _, ev := range s.events {
		if ev.Timestamp.After(since) {
			result = append(result, ev)
		}
	}
	return result
}

// Count returns the total number of stored events.
func (s *Store) Count() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.events)
}

File: custom_components/homestate/facts.py
"""Facts Layer — raw event buffer."""

from __future__ import annotations

from collections import deque
from datetime import datetime

from .models import FactEvent


class FactsStore:
    """Ring buffer for raw HA events."""

    def __init__(self, limit: int = 5000) -> None:
        self._events: deque[FactEvent] = deque(maxlen=limit)

    def record(self, event: FactEvent) -> None:
        self._events.append(event)

    def latest(self, entity_id: str) -> FactEvent | None:
        for event in reversed(self._events):
            if event.entity_id == entity_id:
                return event
        return None

    def recent(self, entity_id: str, n: int = 10) -> list[FactEvent]:
        result = []
        for event in reversed(self._events):
            if event.entity_id == entity_id:
                result.append(event)
                if len(result) >= n:
                    break
        return list(reversed(result))

    def all_since(self, since: datetime) -> list[FactEvent]:
        return [e for e in self._events if e.timestamp > since]

    @property
    def count(self) -> int:
        return len(self._events)
